src/pybind/mgr/crash/module.py

```diff
@@ -0,0 +1,248 @@
+from mgr_module import MgrModule
+import datetime
+import errno
+import json
+
+
+DATEFMT = '%Y-%m-%d %H:%M:%S.%f'
+
+
+class Module(MgrModule):
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+
+    def handle_command(self, inbuf, command):
+        for cmd in self.COMMANDS:
+            if cmd['cmd'].startswith(command['prefix']):
+                handler = cmd['handler']
+                break
+        if handler is None:
+            return errno.EINVAL, '', 'unknown command %s' % command['prefix']
+
+        return handler(self, command, inbuf)
+
+    @staticmethod
+    def validate_crash_metadata(inbuf):
+        # raise any exceptions to caller
+        metadata = json.loads(inbuf)
+        if 'crash_id' not in metadata:
+            raise AttributeError("missing 'crash_id' field")
+        return metadata
+
+    @staticmethod
+    def time_from_string(timestr):
+        # drop the 'Z' timezone indication, it's always UTC
+        timestr = timestr.rstrip('Z')
+        return datetime.datetime.strptime(timestr, DATEFMT)
+
+    # command handlers
+
+    def do_info(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        val = self.get_store(key)
+        if not val:
+            return errno.EINVAL, '', 'crash info: %s not found' % crashid
+        return 0, val, ''
+
+    def do_post(self, cmd, inbuf):
+        try:
+            metadata = self.validate_crash_metadata(inbuf)
+        except Exception as e:
+            return errno.EINVAL, '', 'malformed crash metadata: %s' % e
+
+        crashid = metadata['crash_id']
+        key = 'crash/%s' % crashid
+        # repeated stores of same item are ignored silently
+        if not self.get_store(key):
+            self.set_store(key, inbuf)
+        return 0, '', ''
+
+    def do_ls(self, cmd, inbuf):
+        keys = []
+        for key in self.get_store_prefix('crash/').iterkeys():
+            keys.append(key.replace('crash/', ''))
+        return 0, '\n'.join(keys), ''
+
+    def do_rm(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        if self.get_store(key) is None:
+            return errno.EINVAL, '', 'crash rm: id %s not present' % crashid
+        self.set_store(key, None)       # removes key
+        return 0, '', ''
+
+    def do_prune(self, cmd, inbuf, testmeta=None):
+        now = datetime.datetime.utcnow()
+
+        keep = cmd['keep']
+        try:
+            keep = int(keep)
+        except ValueError:
+            return errno.EINVAL, '', 'keep argument must be integer'
+
+        keeptime = datetime.timedelta(days=keep)
+
+        # testability
+        if testmeta is not None:
```

Comment:
To handling the test mode a bit more neatly, I'd refactor the main logic (selecting what to prune) into a `_prune` method that takes an iterator and returns a list of IDs to remove.  That way you can test that without having these conditional bits in several places.

Reply:
frankly I was surprised at the lack of testing frameworks around this code to begin with.  I agree that would be a bit cleaner, although I still have to select the testing mode somehow.  Shrug.  If you'd really like me to I can refactor it.

Reply:
Yeah, the reason there isn't more of a library of unit tests around the modules is the awkwardness of trying to load them outside of an actual ceph-mgr daemon (hence the selftest commands).

Reply:
...or maybe just remove it in favor of a qa/ test.  This felt icky writing anyway.